### PR TITLE
New linker script and multi-core support

### PR DESCRIPTION
**Pull request by Disasm**
* Linker script was reworked. Now it uses region aliases to relocate sections. This approach makes it possible to build firmware for both FLASH+RAM and RAM-only targets. Memory definitions now supposed to be present in their corresponding crates (e.g. RAM definition in PAC crate, FLASH definition in board support crate).
* Multi-core support was introduced. Cores are parked with `_mp_hook` function and then awakened in platform-dependent way.
* Documentation was updated to reflect new features.
* New crate version: 0.6.0

Depends on: https://github.com/rust-embedded/riscv/pull/28
Closes: https://github.com/rust-embedded/riscv-rt/issues/26

### CODE REVIEW

File: asm.S [hunk 1]
```diff
@@ -71,6 +71,10 @@ _start:
     // Set frame pointer
     add s0, sp, zero
 
+    // Set trap handler
+    la t0, _start_trap
+    csrw mtvec, t0
```
Comment (ilya-epifanov) [inline asm.S]: Beware that not all targets support Privileged Architecture CSRs. 
I wanted to submit a PR gating the `mtvec::write` in `src/lib.rs`. It's not gonna be possible to conditionally compile an '.S'

File: src/lib.rs [hunk 2]
```diff
@@ -253,9 +250,6 @@ pub unsafe extern "C" fn start_rust() -> ! {
 
     // TODO: Enable FPU when available
 
-    // Set mtvec to _start_trap
-    mtvec::write(&_start_trap as *const _ as usize, mtvec::TrapMode::Direct);
```
Comment (ilya-epifanov) [inline src/lib.rs]: Removing this single line allowed me to run Rust on https://github.com/cliffordwolf/picorv32.

File: asm.S (hunk 1 above)
Comment (Disasm) [reply asm.S]: I thought about it. Both `riscv` and `riscv-rt` depend on running with privileged support, so there is no easy way to support `picorv32` *and* complex cores like FU540. Seems like it's better to have separate `picorv32-rt` library.

Comment (ilya-epifanov) [reply asm.S]: Yeah, makes sense. I'll just fork off this repo then after this PR lands

**Comment by Disasm** [comment]
@bradjc Could you review this?

**Comment by laanwj** [comment]
Nice.

**Comment by Disasm** [comment]
@laanwj Do you have any suggestions or we can merge it?

**Comment by Disasm** [comment]
bors try

**Comment by laanwj** [comment]
> @laanwj Do you have any suggestions or we can merge it?

I haven't tried the code, but the changes look good to me. I made similar (but more hacky) changes to succesfully test multi-core on K210. Mainly: set the trap handler for all cores, then enable MIE and MSOFT so that the other core can be woken up using an IPI:
```patch
+    // Set trap handler
+    la t0, _start_trap
+    csrw mtvec, t0
+    // Enable software interrupts
+    csrs mstatus, 8
+    csrs mie, 8
```
Then to invoke the trap handler on the other core:
```rust
    writeln!(stdout, "Generate IPI for core 1 !").unwrap();
    unsafe {
        (*pac::CLINT::ptr()).msip[1].write(|w| w.bits(1));
    }
```

It looks like `mp_hook` would handle specific things such as enabling cross-core software interrupts, and parking the non-main cores until that.

So looks good for merge to me.

**Comment by Disasm** [comment]
@laanwj I thought about handling other cores with a trap handler, but refused to go this way. One of the reasons is that we should decide which core is the main one upon startup. I wanted to avoid this, because I know at least one case where choosing `core 0` as the main one may not work: when you white a firmware for the RV64G cores on FU540. They are numbered 1..4, so you need a different way to choose this "main" core. Linux performs a ["hart lottery"](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/riscv/kernel/head.S#n34) to choose it. However, this approach requires atomics, and I tried to avoid enforcing their presence.

**Comment by Disasm** [comment]
bors r=laanwj
@laanwj Thank you for reviewing this!

**Pull request merged**